lib/src/ast.dart

```diff
@@ -1,5 +1,13 @@
 /// Base interface for queries.
 abstract class Query {
+  const Query(this.tokenStart, this.tokenEnd);
+
+  /// The start position of the token from the input.
+  final int tokenStart;
```

Comment:
This name feels a bit strange, how about `startIndex` instead? also `endIndex`?
Or event better, following the Dart SDK's `Match` class using `start` and `end`, with similar semantics:
https://api.dart.dev/stable/2.17.5/dart-core/Match/end.html

Reply:
I originally had it as `start` and `end` but that clashed with `RangeQuery`. I think `startIndex` and `endIndex` are better names